File: heap_sort.cpp
/*
* @File      :   heap_sort.cpp
* @Date      :   2021/10/06 19:22:06
* @Author    :   Quiser
* @Email     :   qs0714@foxmail.com
*/
#include <iostream>
#include <ctime>
#include <vector>
#define random(a, b) (rand() % (b - a) + a)
using namespace std;

template <class T>
void showvec(vector<T> &v)
{
    for (auto &&i : v)
    {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}

template <class T>
void heapify(vector<T> &tree, int length, int pos)
{
    int c1 = pos * 2 + 1, c2 = pos * 2 + 2, maxpos = pos;
    if (c1 < length && tree[c1] > tree[maxpos])
        maxpos = c1;
    if (c2 < length && tree[c2] > tree[maxpos])
        maxpos = c2;
    if (maxpos != pos)
    {
        swap(tree[pos], tree[maxpos]);
        heapify(tree, length, maxpos);
    }
}

template <class T>
void build_heap(vector<T> &tree, int length)
{
    int finalParent = static_cast<int>((length - 2) / 2);
    for (int i = finalParent; i >= 0; i--)
    {
        heapify(tree, length, i);
    }
}

template <class T>
void heap_sort(vector<T> &tree)
{
    for (int i = tree.size() - 1; i >= 0; i--)
    {
        build_heap(tree, i + 1);
        swap(tree[0], tree[i]);
    }
}

int main()
{
    srand((int)time(NULL));
    cout << "Hello World" << endl;
    vector<int> l = {};
    for (int i = 0; i < 50; i++)
    {
        l.push_back(random(1,100));
    }
    showvec(l);
    heap_sort(l);
    showvec(l);
    return 0;
}

File: heap_sort.py
#!/usr/bin/env python
# _*_ coding: utf-8 _*_

# @File    :   heap_sort.py
# @Author  :   Quiser
# @Email   :   qs0714@foxmail.com
# @Time    :   2021/10/06 19:06:46
from random import randint


def heapify(tree, length, n):
    c1 = n*2+1
    c2 = n*2+2
    maxpos = n
    if c1 < length and tree[c1] > tree[maxpos]:
        maxpos = c1
    if c2 < length and tree[c2] > tree[maxpos]:
        maxpos = c2
    if maxpos != n:
        tree[n], tree[maxpos] = tree[maxpos], tree[n]
        heapify(tree, length, maxpos)


def build_heap(tree, length):
    finalParent = int((length-2)/2)
    for i in range(finalParent, -1, -1):
        heapify(tree, length, i)


def heap_sort(tree):
    for i in range(len(tree), 0, -1):
        build_heap(tree, i)
        tree[0], tree[i-1] = tree[i-1], tree[0]


def main():
    print('hello world')
    l = [randint(1, 200) for _ in range(200)]
    print(l)
    heap_sort(l)
    print(l)


if __name__ == '__main__':
    main()
